### ISSUE TITLE
Introduce metrics

### ISSUE THREAD

**Issue by fKunstner** [fKunstner/optexp#4]
Metrics are currently hardcoded in the problem class.

It would be cleaner to pass a list of metrics to the experiment class to decide what to log.

For this we need an interface to define how metrics work. 

There are a few corner cases that are not clear how to handle yet, in how to specify
- **Inputs** Different metrics need different inputs, eg
    - L2 norm of the weights just needs the model 
    - "Live" Accuracy needs the predictions and targets for the current batch 
    - Full accuracy needs the predictions and targets for the entire dataset
        - Some need the training set, some need the validation set
- **Evaluation frequency** Different metrics need different evaluation frequencies
    - Live metrics can be run each iteration 
    - Entire dataset metrics should not
